Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/PDAnalyst.java

```diff
@@ -669,34 +671,133 @@ public void updateLog(PLog pLog, APMLog apmLog) throws Exception {
+            .collect(Collectors.toSet())
+            .size();
+    }
+
+    private Map<String, Integer> getResourceCountsGroupedByRole() {
+        Map<String, Set<String>> rolesToResourcesMap = new HashMap<>();
+        filteredPLog.getActivityInstances().stream()
+            .filter(activityInstance -> activityInstance.getAttributes().containsKey(Constants.ATT_KEY_RESOURCE))
+            .forEach(activityInstance -> {
+                String role = activityInstance.getAttributeValue(Constants.ATT_KEY_ROLE);
+
+                if (role == null || role.isEmpty()) {
+                    role = SimulationData.DEFAULT_ROLE;
+                }
+                Set<String> resources = rolesToResourcesMap.get(role);
+                if (resources == null) {
+                    resources = new HashSet<>();
+                }
+                resources.add(activityInstance.getResource());
+                rolesToResourcesMap.put(role, resources);
+            });
+
+        return rolesToResourcesMap.entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, stringSetEntry -> stringSetEntry.getValue().size()));
+    }
+
+    private Map<String, String> getRoleForNodeId(BPMNAbstraction bpmnAbstraction) {
+
+        Map<String, String> activityNameToRoleMap = filteredAPMLog.getActivityInstances().stream()
+            .filter(activityInst -> !ObjectUtils.isEmpty(activityInst.getName()))
+            .collect(Collectors.toMap(
+                // key --> activityName
+                ActivityInstance::getName,
+
+                // value --> act
+                activityInst -> {
+                    String role = activityInst.getAttributeValue(Constants.ATT_KEY_ROLE);
+                    return ObjectUtils.isEmpty(role) ? SimulationData.DEFAULT_ROLE : role;
+                },
+
+                /*
+                 * Merge function --> Ignore duplicate keys
+                 * role1 --> The role currently in the map against the key (activity name)
+                 * role2 --> The new role that matches an existing key (activity name)
+                 *
+                 * An activity can be performed by multiple resources, each with the same or different roles
+                 * We will treat the first role of an activity as the primary role for that activity, and ignore
+                 * the rest.
+                 */
+                (role1, role2) -> role1
+
+            ));
+
+        return bpmnAbstraction.getDiagram().getNodes()
+            .stream()
+            .filter(Activity.class::isInstance)
+            .collect(Collectors.toMap(
+                //key --> node Id
+                bpmnNode -> bpmnNode.getId().toString(),
+                //value --> role
+                bpmnNode -> {
+                    String nodeLabel = bpmnNode.getLabel();
+                    return ObjectUtils.isEmpty(nodeLabel)
+                        ? SimulationData.DEFAULT_ROLE : activityNameToRoleMap.get(nodeLabel);
```

Comment:
For safety reason, should be: return ObjectUtils.isEmpty(nodeLabel) || activityNameToRoleMap.containsKey(nodeLabel) ? SimulationData.DEFAULT_ROLE : activityNameToRoleMap.get(nodeLabel).